torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
   return {std::move(payload), std::move(tensors)};
 }
 
+TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage) {
+  tensorpipe::Message tpMessage;
+  std::vector<torch::Tensor> reservedTensors;
+  std::vector<std::vector<uint8_t>> copiedTensors;
+
+  const std::vector<char>& payload = rpcMessage.payload();
+  c10::List<at::Tensor> tensors = cloneSparseTensors(rpcMessage.tensors());
+
+  // Payload
+  tpMessage.data = (uint8_t*)(payload.data());
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = std::to_string(static_cast<int>(rpcMessage.type())) +
+      ":" + std::to_string(static_cast<int>(rpcMessage.id()));
+
+  // Tensors
+  tpMessage.tensors.reserve(tensors.size());
+  for (at::Tensor tensor : tensors) {
+    // Keep original user tensors and cloned sparse tensors
+    reservedTensors.push_back(tensor);
+    tensorpipe::Message::Tensor tpTensor;
+    torch::jit::Pickler pickler(
+        [&](const void* buf, size_t sz) -> size_t {
+          tpTensor.metadata.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
+    pickler.protocol();
+    pickler.pushIValue(tensor);
+    pickler.stop();
+    const auto& tensorDataVec = pickler.tensorData();
+    TORCH_INTERNAL_ASSERT(
+        tensorDataVec.size() == 1, "There should be single pickled tensor");
+    const auto& tensorData = tensorDataVec.front();
+    // Enforce memory copy if tensor is created from torch::from_blob, means
+    // that the tensor doesn't own the memory.
+    if (!tensorData.storageHasDeleter()) {
```

Comment:
Would I be correct if I assume we don't really need the `WriteableTensorData` to check the storage deleter? If so, going through the full pickler seems like an overkill, and the `getWriteableTensorData` seems always copy CUDA tensors to CPU, is this intentional for TensorPipe? 

This is certainly OK for now, I wonder if we should extract out the meta parsing logic from the pickler into a utility function.

Reply:
Regarding to CUDA -> CPU copy, yes. TensorPipe doesn't support CUDA channel yet and need to move to CPU always. Once CUDA copy is enabled in TensorPipe, we would need to either modify pickler or create a separate util to keep CUDA tensors on GPU.
letting tensors go through pickler is mainly to serialize metadata.. And correct me if I'm wrong, pickler shouldn't do memory copy of tensor storage (for CPU tensors)?